neware_parser/DataEngine.py

```diff
@@ -13,75 +10,82 @@ class DataEngine:
     def resistance(
         degradation_model, barcode_count, cyc_grp_dict, cycle_m, cycle_v,
         svit_and_count, filename,
-    ):
+    ) -> dict:
         typ, off, mode = "dchg", 4, "cc"
 
-        keys = make_keys(cyc_grp_dict, typ)
+        protocols = get_protocols(cyc_grp_dict, typ)
         resistances = []
 
         cycles = [x for x in np.arange(0., 6000., 20.)]
         my_cycle = [(cyc - cycle_m) / tf.sqrt(cycle_v) for cyc in cycles]
 
-        for k_count, k in enumerate(keys):
-            target_voltage = cyc_grp_dict[k]["avg_last_cc_voltage"]
-            target_currents = [cyc_grp_dict[k][Key.I_CC_AVG]]
+        for count, protocol in enumerate(protocols):
+            target_voltage = cyc_grp_dict[protocol]["avg_last_cc_voltage"]
+            target_currents = [cyc_grp_dict[protocol][Key.I_CC_AVG]]
 
             test_results = test_single_voltage(
                 my_cycle,
                 target_voltage,
-                cyc_grp_dict[k][Key.I_CC_AVG],
-                cyc_grp_dict[k][Key.I_PREV_END_AVG],
-                cyc_grp_dict[k][Key.V_PREV_END_AVG],
-                cyc_grp_dict[k][Key.V_END_AVG],
+                cyc_grp_dict[protocol][Key.I_CC_AVG],
+                cyc_grp_dict[protocol][Key.I_PREV_END_AVG],
+                cyc_grp_dict[protocol][Key.V_PREV_END_AVG],
+                cyc_grp_dict[protocol][Key.V_END_AVG],
                 target_currents,
                 barcode_count, degradation_model,
                 svit_and_count[Key.SVIT_GRID],
                 svit_and_count[Key.COUNT_MATRIX]
             )
 
             resistances.append(
-                tf.reshape(test_results["pred_R"], shape = [-1]))
-
-        Pickle.dump(filename, (keys, resistances, cycles))
+                tf.reshape(test_results["pred_R"], shape = [-1])
+            )
+        return {
```

Comment:
I don't like this data layout.
if I understand correctly, each protocol has some cycles and some resistances.
but we also have cc_dchg_capacity, cc_chg_capacity, cv_chg_capacity, scale, shift.

So instead you should have one function that computes everything and you should store the data in a way that is compatible with the compiled_dataset format.
For resistance,
I want 
resistance = numpy.array(
                    resistance,
                    dtype = [
                        (Key.N, "f4"),
                        (Key.R, "f4"),
                    ]
                )

scale = numpy.array(
                    scale,
                    dtype = [
                        (Key.N, "f4"),
                        (Key.SCALE, "f4"),
                    ]
                )

Similarly with shift.

For cc_cap and cv_cap,

for protocol in cc_cap.keys():
    cc_cap[protocol] = numpy.array(
                    cc_cap[protocol],
                    dtype = [
                        (Key.N, "f4"),
                        (Key.Q_CC, "f4"),
                    ]
                )

for vectors, such as q, v
for protocol in cc_cap.keys():
    cc_vq[protocol] = numpy.array(
                    cc_vq[protocol],
                    dtype = [
                        (Key.N, "f4"),
                        (Key.V_CC_VEC, "f4", fit_args[Key.V_N_GRID]),
                        (Key.Q_CC_VEC, "f4", fit_args[Key.V_N_GRID]),
                        (Key.MASK_CC_VEC, "f4", fit_args[Key.V_N_GRID]),
                    ]
                )

similarly with cv_vq.

Then you make a dictionary of the results
{"resistance":resistance, "scale":scale, ... "cc_vq":cc_vq}



Here is an example:
for every protocol, call the model and get a list of lists containing [cycle, q_cc_last, q_cv_last] called result, then  
res = numpy.array(
                    result,
                    dtype = [
                        (Key.N, "f4"),
                        (Key.Q_CC_LAST, "f4"),
                        (Key.Q_CV_LAST, "f4"),
                    ]
                )

then
if len(res) > 0:
   cyc_grp_dict[protocol] =res

for the case of the scalars that don't depend on protocol (i.e. resistance, scale, shift),
simply 
res2 = numpy.array(
                    result,
                    dtype = [
                        (Key.N, "f4"),
                        (Key.R, "f4"),
                        (Key.SCALE, "f4"),
                        (Key.SHIFT, "f4"),
                    ]
                )

and then you put these together as:
{'protocol_dependent': cyc_grp_dict, 'protocol_independent': res2}